### ISSUE TITLE
Adding support for rpm packages

### ISSUE THREAD

**Issue by mcallisterjp** [clearlydefined/clearlydefined#156]
Hi,

I'm part of a project that would like to use `clearlydefined` for tracking dependencies expressed in rpm packages, especially from CentOS and [Mariner](https://github.com/microsoft/CBL-Mariner) linux, and would like to contribute rpm support into the code as well as data about specific packages we have used and/or scraped. Of course, we'd like to do this in a way that harmonizes nicely with everything here and so that the `clearlydefined` team are happy maintaining it!

Looking at the equivalent debian situation as well as the `golang`-supporting enhancements clearlydefined/service#765 and clearlydefined/service#871, I think the process for this will be something like:

- [ ] Agree a design how source code/binary packages are defined and relate, and how rpm coordinates work
- [ ] Add support for parsing rpm entries in the `service` repo
- [ ] Add logic for processing dependencies from rpmfiles
- [ ] Add basic support in the `crawler` (along the lines of [this commit](https://github.com/clearlydefined/crawler/commit/ea1618de0de4663b4d9aeecc9cdbc392edb2feba))
- [ ] Document what's done and working and what's not :slightly_smiling_face: 
- [ ] Add sources of data to the `crawler` and/or contribute data directly to the project database for various package sources as needed

Does this sound like basically a good approach to you? Is there anything else we should bear in mind (other than the general contribution guidelines) before starting work?

Thanks,
Pete

**Comment by bduranc** [comment]
Overall, the approach outlined here seems to me like a reasonable one.

Would we support multiple RPM-based Linux distros, or just those listed here? 
What would be the repository location(s) we'd get the packages from? Is there a de-facto location we could use? 

We'll want to carefully consider the full scope, especially if there could be multiple "providers" of such RPM packages (e.g. CentOS, MS/Mariner, Red Hat, etc.) as we should keep the coordinate format simple and consistent with other supported types.

**Comment by qtomlinson** [comment]
Currently deb (and debsrc) packages are supported by ClearlyDefined.  The coordinates looks like [deb/debian/-/debci/2.4_all](https://clearlydefined.io/definitions/deb/debian/-/debci/2.4_all).
It would be nice to keep rpm coordinates consistent for the ease of use from the api perspective.  There is a also [PURL](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#rpm), which can be a good guide.

**Comment by bduranc** [comment]
Sorry for my delay in response.

To respond to your questions, @tofay and @mcallisterjp :

` Is this the right forum to agree the RPM coordinate design?` 
I cannot think of a more appropriate place atm to discuss it.

` Re the "adding a new harvest source", should I open a new issue on the crawler repository to cover that?`
I took a closer look at this issue and the implementation proposed in crawler/[#490](https://github.com/clearlydefined/crawler/issues/490) to familiarize myself.

Overall, from my general (non-technical) perspective the proposal looks good and reasonable. I agree it would be good to start with one provider first (e.g. Mariner), get things comfortable, and then add the others.

From the technical side, I did look briefly into the Mariner package repository and it appears that there are [different directories](https://packages.microsoft.com/cbl-mariner/2.0/prod/base/) for binaries (by arch) and sources (srpms). 

There appears to be a manifest file in XML format (repomd.xml?), which in turn appears to reference SQLite databases which contain the more in-depth information for each package. 

My experience with RPM-based Linux distros is minimal, but I hope the above understanding is at least somewhat correct.

Might it be possible to get a few examples of what different coordinates would look like (particularly for _$revision_)?

It is also important that we are able to resolve both the binaries and their corresponding source (srpms), because end user’s will expect to see the corresponding source for a given binary RPM if its available. But it looks like this is accounted for in your proposed implementation.

Best,

**Comment by bduranc** [comment]
@qtomlinson please add any comments or concerns you might have.

**Comment by nickvidal** [comment]
Is support for rpm still desired?

**Comment by nickvidal** [comment]
Closing for now, feel free to open if interested.
